File: CTCI/Java/DataStructures/SymbolTable.java
package structures;
import structures.LinkedList;

public class SymbolTable<K,T> {

  private Node head;
  private int size = 0;

  private class Node {
    public K key;
    public T data;
    public Node next;
    public Node(K key, T data, Node next) {
      this.key = key; this.data = data; this.next = next;
    }
  }

  public int size() {
    return size;
  }

  public boolean contains(K key) {
    return get(key) != null;
  }

  public T get(K key) {
    for (Node cursor = head; cursor != null; cursor = cursor.next)
      if (key.equals(cursor.key)) return cursor.data;
    return null;
  }

  public void put(K key, T data) {
    if (data == null) {
      delete(key); return;
    }
    for (Node cursor = head; cursor != null; cursor = cursor.next)
      if (key.equals(cursor.key)) {
        cursor.data = data; return;
      }
    head = new Node(key, data, head);
    ++size;
  }

  public void delete(K key) {
    if (head == null) return;
    if (key.equals(head.key)) {
      head = head.next;
      --size;
      return;
    }
    Node previous = head;
    for (Node cursor = head.next; cursor != null; cursor = cursor.next) {
      if (key.equals(cursor.key)) {
        previous.next = cursor.next;
        --size;
        return;
      }
      previous = cursor;
    }
  }

  public LinkedList<K> keys() { // @todo Convert to iterable.
    LinkedList<K> keys = new LinkedList<K>();
    for (Node cursor = head; cursor != null; cursor = cursor.next) {
      keys.appendToTail(cursor.key);
    }
    return keys;
  }

  public static void main(String[] args) {
  }

}

File: CTCI/Go/symbol_table.go
package main

type SymbolTableNode struct {
	next *SymbolTableNode
	data int
	key  int
}

func NewSymbolTableNode(key int, data int) *SymbolTableNode {
	n := new(SymbolTableNode)
	n.data = data
	n.key = key
	return n
}

// SymbolTable (Linked List)
type SymbolTable struct {
	head *SymbolTableNode
	size int
}

func NewSymbolTable() *SymbolTable {
	l := new(SymbolTable)
	return l
}

func (list *SymbolTable) GetTail() *SymbolTableNode {
	if list.head == nil {
		return nil
	}
	var cursor = list.head
	for cursor.next != nil {
		cursor = cursor.next
	}
	return cursor
}

func (list *SymbolTable) Put(key int, data int) {
	n := NewSymbolTableNode(key, data)
	list.AppendNodeToTail(n)
}

func (list *SymbolTable) AppendNodeToTail(node *SymbolTableNode) {
	if list.head == nil {
		list.head = node
	} else {
		list.GetTail().next = node
	}
	list.size++
}

func (list *SymbolTable) Get(key int) int {
	if list.head == nil {
		return -1
	}
	var cursor = list.head
	for ; cursor != nil; cursor = cursor.next {
		if cursor.key == key {
			return cursor.data
		}
	}
	return -1
}

/* Main (Test) */
func main() {
	l := NewSymbolTable()
	l.Put(2, 2)
	println(l.Get(2))
}
